dali/pipeline/graph_descr.cc

```diff
@@ -71,358 +73,444 @@ void CheckOpConstraints(const OpSpec &spec) {
+  auto &left = tensor_nodes_[left_id];
+  auto &right = tensor_nodes_[right_id];
+  // Change ids in producers (there is only one), and on edges
+  {
+    auto &left_prod = Node(left.producer_edge.node);
+    left_prod.children_tensors[left.producer_edge.index] = right_id;
+    left.producer_edge.tensor = right_id;
+    auto &right_prod = Node(right.producer_edge.node);
+    right_prod.children_tensors[right.producer_edge.index] = left_id;
+    right.producer_edge.tensor = left_id;
   }
-
-  // Remove references to this node as a consumer
-  for (int i = 0; i < target.spec.NumInput(); ++i) {
-    auto it = tensor_consumers_.find(target.spec.Input(i));
-    DALI_ENFORCE(it != tensor_consumers_.end(), "Could not find "
-        "consumer entries for tensor, but target node is a consumer.");
-    vector<TensorMeta> &consumer_info = it->second;
-    bool erased = false;
-    for (size_t j = 0; j < consumer_info.size(); ++j) {
-      if (consumer_info[j].node == id) {
-        consumer_info.erase(consumer_info.begin() + j);
-        erased = true;
-        break;
-      }
+  // Change ids in consumers
+  {
+    for (auto &cons_edge : left.consumer_edges) {
+      auto &cons = Node(cons_edge.node);
+      cons.parent_tensors[cons_edge.index] = right_id;
+      cons_edge.tensor = right_id;
+    }
+    for (auto &cons_edge : right.consumer_edges) {
+      auto &cons = Node(cons_edge.node);
+      cons.parent_tensors[cons_edge.index] = left_id;
+      cons_edge.tensor = left_id;
     }
-    DALI_ENFORCE(erased, "Could not find entry for target node as tensor consumer.");
   }
+  // Clean up names mapping
+  tensor_name_to_id_[left.name] = right_id;
+  tensor_name_to_id_[right.name] = left_id;
+  // Swap the actual nodes
+  left.id = right_id;
+  right.id = left_id;
+  std::swap(left, right);
+}
 
-  for (int i = 0; i < this->NumOp(); ++i) {
-    OpNode &node = this->node(i);
-    if (node.id > id) {
-      // Decrement this nodes id to account for
-      // the removal of the node with id `id`.
-      --node.id;
-
-      // Update all of its outputs with the new id
-      for (int j = 0; j < node.spec.NumOutput(); ++j) {
-        auto it = tensor_producers_.find(node.spec.Output(j));
-        DALI_ENFORCE(it != tensor_producers_.end(),
-            "Could not find tensor source entry.");
+void OpGraph::RemoveTensorNode(TensorNodeId id) {
+  DALI_ENFORCE_VALID_INDEX(id, (Index)tensor_nodes_.size());
+  DALI_ENFORCE(tensor_nodes_[id].consumer_edges.empty(),
+               "Removed tensors cannot have any consumers.");
+  // Swap it out
+  for (TensorNodeId i = id + 1; i < static_cast<int>(tensor_nodes_.size()); i++) {
+    // Move from i to i - 1
+    SwapTensorNodes(i, i - 1);
+  }
+  // We remove the last element
+  tensor_nodes_.pop_back();
+  // There is no option to remove from positional array of tensor produced by parent op
+}
 
-        it->second.node = node.id;
+void OpGraph::SwapOpNodes(OpNodeId left_id, OpNodeId right_id) {
```

Comment:
Few things here:
1. Please add unit test
2. Could the two lambdas inside be extracted as free functions? It'd be way clearer, function will be shorter and they could be tested

Reply:
I added the docs. It is tested by Executor and Graph tests.

The lambdas were added to remove confusion and duplication, because previously I had just that implementation repeated two times with swapped names. They really don't make sense out of this scope, and would probably induce confusion.